File: Client/XyMusicForAndroid/domain/src/main/java/com/xymusic/app/feature/playlist/domain/PlaylistUseCases.kt
package com.xymusic.app.feature.playlist.domain

import com.xymusic.app.feature.playlist.domain.model.AddPlaylistTrackCommand
import com.xymusic.app.feature.playlist.domain.model.CreatePlaylistCommand
import com.xymusic.app.feature.playlist.domain.model.PlaylistSort
import com.xymusic.app.feature.playlist.domain.model.RemovePlaylistTrackCommand
import com.xymusic.app.feature.playlist.domain.model.ReorderPlaylistCommand
import com.xymusic.app.feature.playlist.domain.model.UpdatePlaylistCommand
import javax.inject.Inject

class PlaylistUseCases
@Inject
constructor(private val repository: PlaylistRepository) {
    fun playlists() = repository.observePlaylists()

    fun playlist(playlistId: String) = repository.observePlaylist(playlistId)

    suspend fun refreshPlaylists(sort: PlaylistSort = PlaylistSort.UPDATED_DESC) = repository.refreshPlaylists(sort)

    suspend fun refreshPlaylist(playlistId: String) = repository.refreshPlaylist(playlistId)

    suspend fun loadPlaylistPage(playlistId: String, cursor: String? = null) =
        repository.loadPlaylistPage(playlistId, cursor)

    suspend fun create(command: CreatePlaylistCommand) = repository.create(command)

    suspend fun update(command: UpdatePlaylistCommand) = repository.update(command)

    suspend fun delete(playlistId: String, expectedVersion: Long) = repository.delete(playlistId, expectedVersion)

    suspend fun addTrack(command: AddPlaylistTrackCommand) = repository.addTrack(command)

    suspend fun removeTrack(command: RemovePlaylistTrackCommand) = repository.removeTrack(command)

    suspend fun reorder(command: ReorderPlaylistCommand) = repository.reorder(command)
}

File: Client/XyMusicForWin/src/application/use-cases/PlaylistUseCases.ts
import type { PlaylistRepository } from "../ports/PlaylistRepository";
import type { Playlist, PlaylistDetail, PlaylistVisibility } from "../../domain/music";
import type { PlaylistSort } from "../../domain/pagination";

export class PlaylistUseCases {
  constructor(private readonly repository: PlaylistRepository) {}

  list(sort: PlaylistSort, cursor?: string, limit = 50, signal?: AbortSignal) { return this.repository.list(sort, cursor, limit, signal); }
  get(playlistId: string, signal?: AbortSignal) { return this.repository.get(playlistId, signal); }
  getPage(playlistId: string, cursor?: string, limit = 100, signal?: AbortSignal) { return this.repository.getPage(playlistId, cursor, limit, signal); }
  create(name: string, description: string, visibility: PlaylistVisibility) {
    return this.repository.create(name, description, visibility);
  }
  update(playlist: Playlist, changes: { name?: string; description?: string; visibility?: PlaylistVisibility }) {
    return this.repository.update(playlist, changes);
  }
  delete(playlist: Playlist) { return this.repository.delete(playlist); }
  addTrack(playlist: Playlist, trackId: string) { return this.repository.addTrack(playlist, trackId); }
  removeTrack(playlist: PlaylistDetail, entryId: string) { return this.repository.removeTrack(playlist, entryId); }
  reorder(playlist: PlaylistDetail, orderedEntryIds: string[]) { return this.repository.reorder(playlist, orderedEntryIds); }
}
